### ISSUE TITLE
Fix units on dlg_settings

### ISSUE THREAD

**Issue by rpachaly** [FLO-2DSoftware/qgis-flo-2d-plugin#1594]
Improve the if statement to cover all types of units.

https://qgis.org/pyqgis/3.34/core/QgsUnitTypes.html#qgis.core.QgsUnitTypes.DistanceUnit

**Comment by FLO-2DKaren** [comment]
**Note to Users:**  

If you come across this bug, you are using QGIS 3.40.1 and a coordinate system that uses US Feet instead of feet.  You will get an error message like this: 

**WARNING 060319.1654: Unknown map units. Choose a different projection!**

Until Robson fixes this issue, please use QGIS 3.34.

**Comment by rpachaly** [comment]
This is done on https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/compare/fix-units-settings?expand=1

I'll not merge it before your feedback, Karen. Here is a video explaining what I did. 

https://github.com/user-attachments/assets/51953aff-ac68-450b-8495-fbff42341b7a

**Comment by FLO-2DKaren** [comment]
Hi Robson,  I think this is still not working on QGIS 3.40


https://github.com/user-attachments/assets/990cb163-318d-40b1-a57e-663b6a02c33d

**Comment by rpachaly** [comment]
Ok. I'll try to fix this after Jim's webinar.

**Comment by FLO-2DKaren** [comment]
@rpachaly 
Hi Robson,  I was also thinking that you could set a units if you know your units are valid on the map. 
But that might encourage people to select LatLong as their CRS.

**Comment by FLO-2DKaren** [comment]
No Hurry, I"m just getting through some tests while a model finishes running

**Comment by rpachaly** [comment]
Karen, it is fixed. But check this video https://flo-2d.sharefile.com/d-s56da7f9f510e4514b3d8cbac157d7a84 before accepting the pull request. https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/pull/1609

**Comment by FLO-2DKaren** [comment]
Hi Robson,  Wow,  That is a lot of new stuff for QGIS 3.40 units.  So interesting.  I bet it's due to some vertical transformation stuff they will be using.  

I have 1 idea.  Since we don't really care about what units are in the geopackage, we really only want to define them in cont.dat as an on/off switch, we could use a question like this:

The plugin cannot configure the project units for this coordinate system.  Do you want to use International SI or English?  But we only allow that option of the CRS is something other than Degrees.

That way if someone in India makes a project with Indian units, they'll be applied correctly as the "English" method even though they may apply a slightly different ft - m correction.  That doesn't matter to us that the data is a little off.  What matters to us is that the English Metric switch is applied correctly.  Because all we do is read numbers from the topo.dat.  We don't care about how the numbers were created.

And you may not know this but all FLO-2D calculations are done in the English system.  We then convert them back to meters when the *.out files are written.

**Comment by rpachaly** [comment]
That's a great idea. I'll do that in a second.

**Comment by rpachaly** [comment]
![image](https://github.com/user-attachments/assets/7e285f9b-fa52-48fc-a58d-8426775480fc)

**Comment by rpachaly** [comment]
I pushed to that pull request. 

Just letting you know that if you select one of these weird crs, it will default to the WGS84.

**Comment by FLO-2DKaren** [comment]
Tested.

**Comment by FLO-2DKaren** [comment]
Wow!!! So glad I went back and tried this on 3.34.  This is ridiculous.


https://github.com/user-attachments/assets/8f238818-ca5e-4d0e-aeae-c35e3a22ff61

**Comment by FLO-2DKaren** [comment]
Yes very cool.  The only reason I was thinking about dangling elements is for imported projects.  But I agree that it will be difficult to show that.

**Comment by rpachaly** [comment]
Usually when I find an error related to QGIS versions, I use a try/except block. Something like:

try: 
       The new stuff added in newer QGIS versions
except:
      pass or use the old stuff

This is what I did in this case and it seems to be working good. Done and merged on https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/pull/1617

**Comment by FLO-2DKaren** [comment]
Failed Test  Odd CRS test.  What if someone tries to use Yards?

CRS
Kalianpur 1880 / India Zone IIa, the units are **yards**

Sorry I know nobody would choose this system on purpose.  It's just an example that fails the test.

QGIS 3.34 works correctly.  It would not allow me to select  this CRS.  

QGIS 3.40.1 works **incorrectly**.  The plugin asks the user to select the correct units.  I think we should ask to select a different coordinate system.

Yards are not a unit that we can use for grid sizes.

**Comment by FLO-2DKaren** [comment]
**Important Note**   Let's simplify this so we add a selector when units can't be defined, user can enter them.  
Make it a bool selector so that the options are only metric and ft

**Comment by rpachaly** [comment]
So, here is what I did. If the CRS is DistanceFeet or DistanceMeters, proceed. Otherwise, this message shows up

![Image](https://github.com/user-attachments/assets/e630ffb4-2067-415a-96b0-dc9c7dcaf72d)

If yes, the CRS Selector will open again, if no, it will get back to the settings dialog. 

What do you think about the message? If you like it, I'll merge it.

**Comment by FLO-2DKaren** [comment]
Did you take out the code that was looking for "inches", "milimeter"  and such?

**Comment by rpachaly** [comment]
Yes, it only looks for DistanceFeet or DistanceMeters. Otherwise, that message shows up.

**Comment by FLO-2DKaren** [comment]
Does that put us back to the original problem of not reading survey ft correctly?

**Comment by rpachaly** [comment]
Unfortunately yes, but instead of getting an error, the user has the opportunity to select another CRS.

**Comment by rpachaly** [comment]
Ok, that's a great idea. I'm wrapping up the OSM and SSURGO here and when I finish it I'll do it.

**Comment by FLO-2DKaren** [comment]
I was scanning this page and it has new and old systems and all of them have the full word feet or meters.  

https://qgis.org/pyqgis/3.40/core/QgsUnitTypes.html#module-QgsUnitTypes

**Comment by rpachaly** [comment]
Hi Karen, 

So, it is done and tested. It works pretty fine for the US Survey distance units. Here is the code

```
map_units = QgsUnitTypes.toString(self.crs.mapUnits())

if "meters" in map_units:
    self.si_units = True
    mu = "Metric (International System)"
    break  # Exit loop if valid CRS with meters is selected
elif "feet" in map_units:
    self.si_units = False
    mu = "English (Imperial System)"
    break  # Exit loop if valid CRS with feet is selected
else:
    answer = self.uc.customized_question("WARNING 060319.1655",
        "Choose a valid CRS!\n\nFLO-2D only supports coordinate reference systems with distance units in feet or meters.\n\nSelect another CRS?"
    )
    if answer == QMessageBox.Yes:
        continue
    else:
        self.proj_lab.setText("----")
        self.gpkgPathEdit.setText("")
        return
```

It creates a string for the map_units and if meters or feet are in this string, it defines the unit system. Otherwise, it ask the user for a valid CRS. 

Working on 3.34 and 3.40. 

On https://github.com/FLO-2DSoftware/qgis-flo-2d-plugin/pull/1637

**Comment by rpachaly** [comment]
Wait a little. I need to fix this in other parts of the code.

**Comment by rpachaly** [comment]
Done!

**Comment by FLO-2DKaren** [comment]
> ![image](https://github.com/user-attachments/assets/7e285f9b-fa52-48fc-a58d-8426775480fc)

Hi Robson,

What happened to this?  I thought it was a good idea?  Let's reimplement it for the users who don't have a language in our keyword lists.
